### DISCUSSION TITLE
Custom gRPC input

### DISCUSSION THREAD

**Discussion by sagikazarmark** [redpanda-data/connect#2380] [General]
I'm trying to implement a custom gRPC input as a plugin (something similar to the HTTP server input).

I can see that the HTTP server input implementation is different from the public plugin interface. Is there a plugin implementation out there that does something similar?

Given most of the transaction handling code (that I could find in the HTTP server input) is internal, would it make sense to copy some of those? Or is there a better way?

I'm guessing the v4 plugin interface replaced some of that with ack functions, but I'm not entirely sure what would be the best way to implement something like this.

Any pointers in the right direction are appreciated! Thanks!

**Comment by mfamador** [comment]
Hey @sagikazarmark. 
I've made this example while ago, you might want to use it as a starting point.
https://github.com/mfamador/benthos-input-grpc/tree/main

**Reply by sagikazarmark** [reply]
@mfamador appreciate the reply!

I could easily be wrong here, but it seems to me that the way your example is implemented doesn't guarantee that an error is returned to the right caller (ie. in case of multiple parallel gRPC calls each call will block on reading from the error channel, but AFAIK it is not guaranteed they will receive from the channel in the order they started listening)

Am I wrong in my assumption?

In any case, I came up with something similar myself, so it seems I'm not on a totally wrong path here.

Thanks!
